core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1106,6 +1106,13 @@ class Log(@volatile private var _dir: File,
         // they are valid, insert them in the log
         lock synchronized {
           checkIfMemoryMappedBufferClosed()
+
+          // check for offline log dir in case a retry following an IOException happens before the log dir
```

Comment:
Hmm.. In case there is an IOException on an append, we will release the lock and fail the log dir in `maybeHandleIOException`. There is a window for another append to sneak by. It looks like it should be possible to pull `maybeHandleIOException` into the locked section here since `analyzeAndValidateRecords` does not do any IO.